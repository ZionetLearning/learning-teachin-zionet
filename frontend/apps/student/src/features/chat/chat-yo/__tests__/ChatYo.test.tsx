import { render, screen, fireEvent } from "@testing-library/react";
import { vi, describe, it, expect } from "vitest";
import { ChatYo } from "../";

// spy-able sendMessage
const sendMessageMock = vi.fn();

vi.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (k: string) => k }),
}));

// hook mock
vi.mock("@student/hooks", () => ({
  useChat: () => ({
    sendMessage: sendMessageMock,
    loading: false,
    messages: [
      { text: "Hello", position: "right", date: new Date() },
      { text: "Hi there!", position: "left", date: new Date() },
    ],
  }),
}));

// mock ChatUi to render a minimal input and button that call the passed props
vi.mock("@student/components", () => ({
  ReactChatElements: ({
    messages,
    loading,
    value,
    onChange,
    handleSendMessage,
  }: {
    messages: Array<{ text: string }>;
    loading: boolean;
    value: string;
    onChange: (v: string) => void;
    handleSendMessage: () => void;
  }) => (
    <div>
      <ul>
        {messages.map((m, i) => (
          <li key={i}>{m.text}</li>
        ))}
      </ul>
      {loading && <div>Loading…</div>}
      <input
        placeholder="Type a message..."
        value={value}
        onChange={(e) => onChange((e.target as HTMLInputElement).value)}
      />
      <button type="button" onClick={handleSendMessage}>
        ↑
      </button>
    </div>
  ),
}));

describe("ChatYo", () => {
  it("types, sends, clears input, and calls sendMessage", () => {
    sendMessageMock.mockReset();

    render(<ChatYo />);

    // sees existing messages
    expect(screen.getByText("Hello")).toBeInTheDocument();
    expect(screen.getByText("Hi there!")).toBeInTheDocument();

    const input = screen.getByPlaceholderText(
      "Type a message...",
    ) as HTMLInputElement;

    fireEvent.change(input, { target: { value: "Test message" } });
    expect(input.value).toBe("Test message");

    fireEvent.click(screen.getByRole("button", { name: "↑" }));

    expect(sendMessageMock).toHaveBeenCalledTimes(1);
    expect(sendMessageMock).toHaveBeenCalledWith("Test message");
    expect(input.value).toBe(""); // cleared by ChatYo after send
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<ChatYo />);
    expect(asFragment()).toMatchSnapshot();
  });
});
