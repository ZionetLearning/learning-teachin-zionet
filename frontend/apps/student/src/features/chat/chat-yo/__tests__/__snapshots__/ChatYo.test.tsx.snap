// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatYo > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="chatWrapper-0-2-1"
    data-testid="chat-yo-page"
  >
    <div>
      <ul>
        <li>
          Hello
        </li>
        <li>
          Hi there!
        </li>
      </ul>
      <input
        placeholder="Type a message..."
        value=""
      />
      <button
        type="button"
      >
        â†‘
      </button>
    </div>
  </div>
</DocumentFragment>
`;
